File: frontend/lib/library/court-levels.ts
// Canonical `cases.court_level` vocabulary for the frontend — the TS mirror of
// `agents/deep_search_v4/shared/court_levels.py`. Pure data + string helpers, no
// data fetching.
//
// The column has THREE values in prod (2026-07-24):
//   first_instance  23,932   appeal  6,474   supreme  125
//
// The Python module exists because that third value was silently collapsed in
// FOUR independent places, each written as a two-branch ternary — every
// supreme-court ruling was relabelled ابتدائي downstream. Import from here
// instead of re-deriving the mapping locally; a two-branch conditional over a
// three-value column is the bug, and it reappears every time someone re-writes
// it inline.
//
// DISPLAY RULE: the backend ships a rendered `court_level_label` on every
// judgment payload and OWNS the display vocabulary. `courtLevelLabel()` prefers
// that label and only falls back to the map below when it's missing — so a
// future fourth value shows the backend's wording, not a stale local guess.

import type { CourtLevel } from "@/types/library";

/** Canonical enum values, ordered by judicial seniority (ascending). */
export const COURT_LEVELS: readonly CourtLevel[] = [
  "first_instance",
  "appeal",
  "supreme",
] as const;

/** enum → Arabic display label. */
export const COURT_LEVEL_AR: Record<CourtLevel, string> = {
  first_instance: "ابتدائي",
  appeal: "استئناف",
  supreme: "المحكمة العليا",
};

/** True when `raw` is one of the three canonical values. */
export function isCourtLevel(raw: string | null | undefined): raw is CourtLevel {
  return !!raw && (COURT_LEVELS as readonly string[]).includes(raw);
}

/**
 * Arabic label for a court level. Prefers the backend's own `court_level_label`
 * when supplied; falls back to the local map; returns `""` for an unrecognised
 * value with no label — printing ابتدائي for an unknown level would assert
 * something false about a judgment, so display paths fail to silence, not to a
 * default.
 */
export function courtLevelLabel(
  raw: string | null | undefined,
  backendLabel?: string | null,
): string {
  const fromBackend = backendLabel?.trim();
  if (fromBackend) return fromBackend;
  return isCourtLevel(raw) ? COURT_LEVEL_AR[raw] : "";
}

/** One option in the hub's court-level filter row. `value: ""` = «الكل». */
export interface CourtLevelFilterOption {
  value: string;
  label: string;
}

/**
 * The hub filter row, in seniority order behind «الكل». Values are the raw enum
 * tokens the backend's `court_level` query param expects.
 */
export const COURT_LEVEL_FILTERS: readonly CourtLevelFilterOption[] = [
  { value: "", label: "الكل" },
  ...COURT_LEVELS.map((level) => ({
    value: level,
    label: COURT_LEVEL_AR[level],
  })),
] as const;

File: agents/deep_search_v4/shared/court_levels.py
"""Canonical `cases.court_level` vocabulary — one home, no per-module copies.

`cases.court_level` has **three** values in prod (2026-07-24):

    first_instance  23,932
    appeal           6,474
    supreme            125   <-- the one every ad-hoc copy of this logic dropped

This module exists because that third value was silently collapsed in FOUR
independent places, each written as a two-branch ternary or a two-alternative
regex:

    case_search/unfold_ura.py   `"appeal" if raw == "appeal" else "first_instance"`
    case_search/reranker.py     `r"\\((ابتدائي|استئناف)\\)"` + two-value ternary
    case_search/search.py       `"استئناف" if lvl == "appeal" else "ابتدائي"`
    aggregator/preprocessor.py  (new Arabic label map)

Every supreme-court ruling was therefore labelled ابتدائي / `first_instance`
downstream. Import from here instead of writing a fourth copy — a two-branch
conditional over a three-value column is the bug, and it reappears every time
someone re-derives it locally.

See `.claude/plans/case_topics_loop.md` §8 (decision D5: court level is
informational only — it is never a retrieval boost or a relevance signal).
"""
from __future__ import annotations

# Canonical enum values, ordered by judicial seniority (ascending).
CASE_COURT_LEVELS: tuple[str, ...] = ("first_instance", "appeal", "supreme")

# Fallback for NULL / empty / unrecognised input. `cases.court_level` is NOT
# NULL in the schema, so this only fires on a hand-built row or a future value.
DEFAULT_COURT_LEVEL = "first_instance"

# enum -> Arabic display label (prompts, reranker markdown, aggregator refs).
COURT_LEVEL_AR: dict[str, str] = {
    "first_instance": "ابتدائي",
    "appeal": "استئناف",
    "supreme": "عليا",
}

# Arabic label -> enum. Used when parsing a rendered header back to the enum
# (the legacy markdown reranker path). Keys are exactly COURT_LEVEL_AR values.
COURT_LEVEL_FROM_AR: dict[str, str] = {ar: en for en, ar in COURT_LEVEL_AR.items()}


def normalize_court_level(raw: str | None) -> str:
    """Coerce a raw `cases.court_level` value to one of :data:`CASE_COURT_LEVELS`.

    Unknown / NULL / empty input returns :data:`DEFAULT_COURT_LEVEL`. Never
    raises — this sits on the retrieval hot path and a surprise value must not
    take down a search.
    """
    value = (raw or "").strip()
    return value if value in CASE_COURT_LEVELS else DEFAULT_COURT_LEVEL


def court_level_ar(raw: str | None, *, strict: bool = False) -> str:
    """Arabic label for a court level.

    Args:
        raw: the enum value (or None).
        strict: when True, an unrecognised value yields ``""`` instead of the
            default label. Use this for **display** paths, where silently
            printing ابتدائي for an unknown level would assert something false;
            leave it False on paths that need a label no matter what.
    """
    value = (raw or "").strip()
    if value in COURT_LEVEL_AR:
        return COURT_LEVEL_AR[value]
    return "" if strict else COURT_LEVEL_AR[DEFAULT_COURT_LEVEL]


__all__ = [
    "CASE_COURT_LEVELS",
    "COURT_LEVEL_AR",
    "COURT_LEVEL_FROM_AR",
    "DEFAULT_COURT_LEVEL",
    "court_level_ar",
    "normalize_court_level",
]
